### DISCUSSION TITLE
Apply UI changes everytime for an application?

### DISCUSSION THREAD

**Discussion by sopgenorth** [snoopwpf/snoopwpf#441] [Q&A] [answered]
I want to modify the UI of an application that I don't have the source code for. I can use snoop though to change UI elements to what I want and then use the application, but re-applying these UI tree element changes every time is a bit cumbersome...

So, my thought was if there were some way to save the UI element changes that I made, and then re-apply them every time the application launches. Is this possible to achieve with snoop? Or, does anyone know of an alternative method for achieving this goal?

**Comment by batzen** [comment, marked as answer]
There is no way to do this with Snoop itself.
Re-Applying would require finding the exact controls the changes would have to be applied to.
That might work quite well if the target application provides enough names/uids for controls to find them again.

What you can do:
- Take a copy of the injector assemblies/executables of Snoop
- Write a tool/script that identifies the PID of the target process
- Create a .NET assembly that has a class with a static method that accepts one string parameter
- Call the `Snoop.InjectorLauncher,.exe` with your custom built commandline
  - The commandline has to contain `-TargetPID XYZ -Assembly PathToYourOwnNETAssembly -ClassName FullyQualifiedClassNameInYourAssembly -MethodName MethodNameOnYourClass -SettingsFile OptionalPathToAFileThatContainsInformationYouWantToProvideToYourMethod`
- When everything worked the code from your assembly now runs in the target process and you can manipulate whatever you want. To get all windows currently open you can use
`Application.Current.Windows`
Be aware that your code might run in a different thread than the main UI thread.
To switch threads you can run your code with `Application.Current.Dispatcher.Invoke`.

**Reply by batzen** [reply]
As a small sample you can have a look at https://github.com/snoopwpf/snoopwpf/tree/develop/TestHarnesses/SimpleUIAutomation.Application and https://github.com/snoopwpf/snoopwpf/tree/develop/TestHarnesses/SimpleUIAutomation

**Reply by sopgenorth** [reply]
That will do! Thank you for this suggestion
